shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Generator/Utxo.hs

```diff
@@ -136,174 +133,85 @@ genTx
-          -- deterministically for each script. Varying the script is possible though.
-
-          let wits =
-                mkTxWits
-                  ksIndexedPaymentKeys
-                  ksIndexedStakingKeys
-                  (hashAnnotated txBody)
-                  ( spendWitnesses
-                      ++ wdrlWitnesses
-                      ++ certWitnesses
-                      ++ updateWitnesses
-                      ++ genesisWitnesses
-                      ++ genesisDelegationWitnesses
-                  )
-                  multiSig
-
-          let metadata = SNothing -- TODO generate metadata
-
-          -- calculate real fees of witnesses transaction
-          let minimalFees = minfeeBound pparams (Tx txBody wits metadata)
-
-          -- discard generated transaction if the balance cannot cover the fees
-          if minimalFees > balance_
-            then QC.discard
-            else do
-              -- update model transaction with real fees and outputs
-              let (fees', outputs') =
-                    calcOutputsFromBalance balance_ recipientAddrs minimalFees
-                  txBody' =
-                    txBody
-                      { _txfee = fees',
-                        _outputs = outputs'
-                      }
-                  wits' =
-                    mkTxWits
-                      ksIndexedPaymentKeys
-                      ksIndexedStakingKeys
-                      (hashAnnotated txBody')
-                      ( spendWitnesses
-                          ++ wdrlWitnesses
-                          ++ certWitnesses
-                          ++ updateWitnesses
-                          ++ genesisWitnesses
-                          ++ genesisDelegationWitnesses
-                      )
-                      multiSig
-
-              pure (Tx txBody' wits' metadata)
+      let metadata = SNothing -- TODO generate metadata
+      ttl <- genTimeToLive
+      -------------------------------------------------------------------------
+      -- Gather Key Witnesses and Scripts, prepare a constructor for Tx Wits
+      -------------------------------------------------------------------------
+      let wits = spendWits ++ wdrlWits ++ certWits ++ updateWits
+          scripts = mkScriptWits spendScripts (certScripts ++ wdrlScripts)
+          mkTxWits' = mkTxWits ksIndexedPaymentKeys ksIndexedStakingKeys wits scripts . hashAnnotated
+      -------------------------------------------------------------------------
+      -- SpendingBalance, Output Addresses (including some Pointer addresses)
+      -- and a Outputs builder that distributes the given balance over addresses.
+      -------------------------------------------------------------------------
+      let spendingBalance = assertPositive $ spendingBalanceUtxo + (sum (snd <$> wdrls)) - deposits + refunds
+          assertPositive b = if b > 0 then b else error ("genTx: zero or negative balance " <> show b)
```

Comment:
@mrBliss there are now 2 errors in `genTx` that replace what was previously `Gen.discard`. The consensus tests that use `genTx` checks for discards, would it be simple enough for now to deal with this change in the consensus layer?

The plan (with @nfrisby) is to in the next round, instead of error/discard, return a callback function/continuation instead.

Reply:
> @mrBliss there are now 2 errors in `genTx` that replace what was previously `Gen.discard`. The consensus tests that use `genTx` checks for discards, would it be simple enough for now to deal with this change in the consensus layer?

I would think so, both `error` and `discard` are exceptions.

> 
> The plan (with @nfrisby) is to in the next round, instead of error/discard, return a callback function/continuation instead.

I see.

BTW, these `error`s, could they be avoided by more cleverly picking a value beforehand, i.e., one that doesn't have a negative balance? That's why I typically do.

Reply:
Hi @uroboros, sorry I was unexpectedly unavailable today. I replied to your Slack message and am copying that here.

https://github.com/input-output-hk/ouroboros-network/blob/f1c094627fc45e6a82f9a9ad8ee722e5a6b0e239/ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs#L653-L674

^^^ That's the only piece of code I'm aware of that relies on the specifics of the error in this case.

So if the generator changes to throw `FooException` instead of `QC.discard`, we'd just need to change the `try` and/or `isDiscard` in that `where` clause to only catch the `FooException`.